Issue #6329 Fix xml config problems

janbartel opened the pull request:
Closes #6392 

The original problem that was reported in #6329 was the configuration for `Server.setStopAtShutdown` had been mistakenly  changed from a `default` of `true` to `false`, prompting a full review of the changes from the relevant commit (https://github.com/eclipse/jetty.project/commit/343cf732b9a5aacd4c84fcb937a4b1624a1a0afd). This PR puts back some other changes that look like mistakes, but can reviewers @sbordet @gregw and @joakime check carefully if these changes are indeed needed. I've simplified the session xml configs to put the defaults into code rather than xml where possible.

janbartel committed:
Issue #6329 Fix xml config problems

Signed-off-by: Jan Bartel <janb@webtide.com>

joakime commented on jetty-server/src/main/java/org/eclipse/jetty/server/Server.java:
```diff
@@ -78,7 +78,7 @@
     private final List<Connector> _connectors = new CopyOnWriteArrayList<>();
     private SessionIdManager _sessionIdManager;
-    private boolean _stopAtShutdown;
+    private boolean _stopAtShutdown = true;
```
This will be a regression for embedded Jetty users that expect it to be false.

The true setting is only relevant for users of jetty-home / jetty-maven-plugin.
This default should be declared in the `server.mod` and code for jetty-maven-plugin.

janbartel replied:
I really don't like the default being different .... but I understand that would be a change for embedded users. I'll make the default back to `true` for xml.

joakime reviewed (changes requested):
The default for Server.stopAtShutdown should remain false (for embedded users).

janbartel committed:
Issue #6329 Put default stopAtShutdown=false

Signed-off-by: Jan Bartel <janb@webtide.com>

gregw commented on jetty-server/src/main/config/etc/jetty.xml:
```diff
@@ -66,8 +66,8 @@
-      <Set name="sendDateHeader" property="jetty.httpConfig.sendDateHeader"/>
-      <Set name="headerCacheSize" property="jetty.httpConfig.headerCacheSize"/>
+      <Set name="sendDateHeader"><Property name="jetty.httpConfig.sendDateHeader" default="false"/></Set>
+      <Set name="headerCacheSize"><Property name="jetty.httpConfig.headerCacheSize" default="4096"/></Set>
```
I think we can change the embedded default for headerCacheSize.
I'm tempted to say the same for sendDateHeader.... but I guess we'd better not.

janbartel replied:
Removed default for headerCacheSize, used code defaults instead.

gregw commented on jetty-http2/http2-server/src/main/config/etc/jetty-http2c.xml:
```diff
@@ -6,8 +6,8 @@
         <Arg name="config"><Ref refid="httpConfig"/></Arg>
-        <Set name="maxConcurrentStreams" property="jetty.http2c.maxConcurrentStreams"/>
-        <Set name="initialStreamRecvWindow" property="jetty.http2c.initialStreamRecvWindow"/>
+        <Set name="maxConcurrentStreams"><Property name="jetty.http2c.maxConcurrentStreams" default="1024"/></Set>
+        <Set name="initialStreamRecvWindow"><Property name="jetty.http2c.initialStreamRecvWindow" default="65535"/></Set>
```
@sbordet Can we change the defaults in the code so we don't need defaults in the XML as well?

sbordet replied:
@gregw we should change these default to match those in the code, as these defaults are quite wrong.

gregw replied:
@sbordet I think the idea is to not have defaults in the XML if at all possible.  If the defaults in the code are correct, then we should just change these to the new form of the XML that only sets if the property is set.

janbartel replied:
Removed defaults from xml in favour of using code defaults instead.

gregw commented on jetty-gcloud/jetty-gcloud-session-manager/src/main/config-template/etc/sessions/gcloud/session-store.xml:
```diff
@@ -10,11 +10,11 @@
        <Set name="maxRetries"><Property name="jetty.session.gcloud.maxRetries" default="5"/></Set>
-       <Set name="backoffMs" property="jetty.session.gcloud.backoffMs"/>
-       <Set name="namespace" property="jetty.session.gcloud.namespace"/>
+       <Set name="backoffMs"><Property name="jetty.session.gcloud.backoffMs" default="1000"/></Set>
```
Can we change this default in the code?  I think changing numeric defaults is OK, as it doesn't change behaviour, just when the behaviour happens.

janbartel replied:
Actually I did default it in the code, I just forgot to remove it from the xml. I'll do that.

gregw commented on jetty-gcloud/jetty-gcloud-session-manager/src/main/config-template/etc/sessions/gcloud/session-store.xml:
```diff
@@ -10,11 +10,11 @@
-       <Set name="gracePeriodSec"><Property name="jetty.session.gracePeriod.seconds" default="3600" /></Set>
+       <Set name="gracePeriodSec" property="jetty.session.gracePeriod.seconds"/>
        <Set name="savePeriodSec" property="jetty.session.savePeriod.seconds"/>
        <Set name="maxRetries"><Property name="jetty.session.gcloud.maxRetries" default="5"/></Set>
```
Can we default this in the code as well?

gregw commented on jetty-unixsocket/jetty-unixsocket-server/src/main/config-template/etc/jetty-unixsocket-http2c.xml:
```diff
@@ -6,8 +6,8 @@
         <Arg name="config"><Ref refid="unixSocketHttpConfig"/></Arg>
-        <Set name="maxConcurrentStreams" property="jetty.http2c.maxConcurrentStreams"/>
-        <Set name="initialStreamRecvWindow" property="jetty.http2c.initialStreamRecvWindow"/>
+        <Set name="maxConcurrentStreams"><Property name="jetty.http2c.maxConcurrentStreams" default="1024"/></Set>
+        <Set name="initialStreamRecvWindow"><Property name="jetty.http2c.initialStreamRecvWindow" default="65535"/></Set>
```
@sbordet ditto - can these be defaulted just in the code?

sbordet replied:
Same.

janbartel replied:
Removed xml defaults in favour of using code defaults instead.

sbordet commented:
@janbartel every change in XML config files requires likely a change in the `[ini-template] section of the correspondent `*.mod` file.

sbordet reviewed (changes requested):
Every change in XML config files requires likely a change in the `[ini-template]` section of the correspondent `*.mod` file.

janbartel commented:
@sbordet the changes I have made do not require any changes in the corresponding `.mod` files - I checked them all and if a value is specified in the `.xml`, it is the same as the (commented out!) value in the `.mod` file and the pre https://github.com/eclipse/jetty.project/commit/343cf732b9a5aacd4c84fcb937a4b1624a1a0afd commit values.

In many cases, the commented out  value in the `.mod ` file or the value supplied in the `.xml` file is _different_ to the defaults in the code. Can you and @gregw please give some guidance on which values can be changed in the code vs which need to remain defaulted only in the `.xml`?

janbartel committed:
Issue #6329 Remove unneeded gcloud sessiondatastore defaults

gregw commented:
@janbartel I'll defer my re-review until @sbordet comments

joakime commented on jetty-gcloud/jetty-gcloud-session-manager/src/main/java/org/eclipse/jetty/gcloud/session/GCloudSessionDataStore.java:
```diff
@@ -65,7 +66,7 @@
     protected EntityDataModel _model;
     protected boolean _modelProvided;
-    private String _namespace;
+    private String _namespace = DEFAULT_NAMESPACE;
```
This is an embedded mode change, it used to return null here, now it returns empty string.

janbartel replied:
No, if the namespace is null at startup the GCloudDataStore uses "".

joakime commented on jetty-gcloud/jetty-gcloud-session-manager/src/main/java/org/eclipse/jetty/gcloud/session/GCloudSessionDataStoreFactory.java:
```diff
@@ -22,9 +22,9 @@
-    private int _backoffMs;
+    private String _namespace = GCloudSessionDataStore.DEFAULT_NAMESPACE;
+    private int _maxRetries = GCloudSessionDataStore.DEFAULT_MAX_RETRIES;
+    private int _backoffMs = GCloudSessionDataStore.DEFAULT_BACKOFF_MS;
```
These 3 are embedded mode default changes.
These used to be (in embedded mode) null, 0, 0.

janbartel replied:
See above

joakime commented on jetty-server/src/main/java/org/eclipse/jetty/server/session/AbstractSessionDataStoreFactory.java:
```diff
@@ -19,8 +19,8 @@
-    int _gracePeriodSec;
-    int _savePeriodSec;
+    int _gracePeriodSec = AbstractSessionDataStore.DEFAULT_GRACE_PERIOD_SEC;
+    int _savePeriodSec = AbstractSessionDataStore.DEFAULT_SAVE_PERIOD_SEC;
```
These are embedded mode changes.
It used to be 0 for each.

janbartel replied:
See above

joakime reviewed (commented):
There are still changes for embedded mode users.

sbordet commented on jetty-server/src/main/config/etc/jetty.xml:
```diff
@@ -105,7 +105,7 @@
     <!-- extra server options                                        -->
     <!-- =========================================================== -->
-    <Set name="stopAtShutdown" property="jetty.server.stopAtShutdown"/>
+    <Set name="stopAtShutdown"><Property name="jetty.server.stopAtShutdown" default="true"/></Set>
```
Does not this require a test with the `JettyHomeTester`?

janbartel replied:
I tried, but it's just not possible with the current setup. When `destroy() `is called on the remote jetty process the IO pipes to that process seem to be immediately closed, so we never see the appropriate `"Stopped Server"` log message (although it is generated, as running the distro at the command line attests). I think this is a much bigger issue, namely how do we test each and every module to ensure correctness, so I've opened a separate issue: #6373

janbartel committed:
Issue #6329 Changes post review

Signed-off-by: Jan Bartel <janb@webtide.com>

janbartel commented:
@gregw @sbordet  I have removed the defaults set in xml in favour of using the defaults in the code as discussed. Where appropriate I have changed the .mod files to also use the same values as the code defaults. Please rereview.

sbordet commented on jetty-gcloud/jetty-gcloud-session-manager/src/main/config-template/etc/sessions/gcloud/session-store.xml:
```diff
@@ -44,6 +44,9 @@
               </Set>
+              <Set name="lastSaved">
+                <Property name="jetty.session.gcloud.model.lastSaved" default="lastSaved" />
+              </Set>
```
What's the rationale to have some as `property` attribute and some as `Property` elements?
The code and the `*.mod` `[ini-template]` are consistent, so I guess they can all be `property` attributes?

janbartel replied:
The `Property` element allows a default value to be set, whereas the `property` attribute does not.

gregw replied:
@sbordet specifically, sometime we "want" to have different defaults for embedded usage (default in the code) vs deployed usage (default in the XML).    I'm not sure we should create any more situations where the default is different depending on the run mode, but we do have several historic cases.     We did originally convert some of them to `property` elements (ie use only the default from the code), but that caused complaints.   So we need to use `Property` element whenever we want the xml to always set a value, even if the property is not set.

janbartel replied:
I've converted all the `<Property>` elements to `property` attributes as the `defaults` are the same as the code (and the corresponding `.mod` file).

sbordet commented on jetty-hazelcast/src/main/config/etc/sessions/hazelcast/remote.xml:
```diff
@@ -10,21 +10,11 @@
+        <Set name="savePeriodSec" property="jetty.session.savePeriod.seconds"/>
         <Set name="onlyClient">
           <Property name="jetty.session.hazelcast.onlyClient" default="true" />
         </Set>
```
Why was this kept as `Property`? Is it correct that embedded sees `false`, but standalone server sees `true`?

janbartel replied:
This is a `Property` because we are setting the `default` value. As the same HazelcastSessionDataStoreFactory is used in both the `remote.xml `and `default.xml`, we cannot default to the value in the class, but must set this field explicitly to match either the remote or embedded case.  The other alternative is to make it a `property`, and remove the default and trust that the value in the .`mod/.ini `file is not commented out/correct.

sbordet commented on jetty-rewrite/src/main/config/etc/jetty-rewrite-customizer.xml:
```diff
@@ -9,7 +9,7 @@
         <Set name="rewriteRequestURI" property="jetty.rewrite.rewriteRequestURI"/>
         <Set name="rewritePathInfo" property="jetty.rewrite.rewritePathInfo"/>
-        <Set name="originalPathAttribute" property="jetty.rewrite.originalPathAttribute"/>
+        <Set name="originalPathAttribute"><Property name="jetty.rewrite.originalPathAttribute" default="requestedPath"/></Set>
```
Seems like a breaking change?

janbartel replied:
It was this way before the commit that changed the xml that kicked off this PR: https://github.com/eclipse/jetty.project/commit/343cf732b9a5aacd4c84fcb937a4b1624a1a0afd

gregw replied:
I think in this case  we can make this one a `property` as I see no reason for the embedded default to be different from the deployed default.

janbartel replied:
So to be crystal clear on this:   `RewriteCustomer` has a default of `null` for `_originalPathAttribute`. The `rewrite-customizer.mod` file has a _commented out_ value for `jetty.rewrite.originalPathAttribute=requestedPath`. Before commit https://github.com/eclipse/jetty.project/commit/343cf732b9a5aacd4c84fcb937a4b1624a1a0afd the xml defaulted it to `requestedPath`. So this PR is simply putting things back the way they were. @gregw and @sbordet do either of you believe the value should now be defaulted _differently_???? If so, to what?

sbordet replied:
This should be restored to the same behavior it had for all the Jetty 10 versions.
The default value of `null` means to not store the originalPath as a request attribute, which is a valid behavior and has been so for all Jetty 10 versions.

sbordet commented on jetty-rewrite/src/main/config/etc/jetty-rewrite.xml:
```diff
@@ -10,7 +10,7 @@
         <Set name="rewriteRequestURI" property="jetty.rewrite.rewriteRequestURI" />
         <Set name="rewritePathInfo" property="jetty.rewrite.rewritePathInfo" />
-        <Set name="originalPathAttribute" property="jetty.rewrite.originalPathAttribute" />
+        <Set name="originalPathAttribute"><Property name="jetty.rewrite.originalPathAttribute" default="requestedPath"/></Set>
```
Ditto.

janbartel replied:
See above comment.

gregw replied:
ditto

janbartel replied:
Both @sbordet and @gregw see above comment.

sbordet commented on jetty-server/src/main/config/etc/jetty-threadlimit.xml:
```diff
@@ -10,7 +10,7 @@
       <New id="ThreadLimitHandler" class="org.eclipse.jetty.server.handler.ThreadLimitHandler">
         <Arg name="forwardedHeader"><Property name="jetty.threadlimit.forwardedHeader"/></Arg> 
-        <Set name="enabled" property="jetty.threadlimit.enabled"/> 
+        <Set name="enabled"><Property name="jetty.threadlimit.enabled" default="true"/></Set>
```
Was not enabled before, why was changed to enabled?

janbartel replied:
Because if its not enabled it's not doing anything, and the value previously was `true`, so I've reinstated it. The value in the class is `false`. Again, see commit https://github.com/eclipse/jetty.project/commit/343cf732b9a5aacd4c84fcb937a4b1624a1a0afd

sbordet replied:
Again, breaks the behavior of all Jetty 10 releases.
If it's a wrong value, let's open a new issue rather than doing it here where it gets lost in other changes.

sbordet commented on jetty-server/src/main/config/etc/jetty.xml:
```diff
@@ -66,7 +66,7 @@
       <Set name="responseHeaderSize" property="jetty.httpConfig.responseHeaderSize"/>
       <Set name="sendServerVersion" property="jetty.httpConfig.sendServerVersion"/>
-      <Set name="sendDateHeader" property="jetty.httpConfig.sendDateHeader"/>
+      <Set name="sendDateHeader"><Property name="jetty.httpConfig.sendDateHeader" default="false"/></Set>
```
Another behavior change? Why?

janbartel replied:
@sbordet I don't think you've understood the fundamental of this PR :)  The whole point is that commit https://github.com/eclipse/jetty.project/commit/343cf732b9a5aacd4c84fcb937a4b1624a1a0afd accidentally changed the configuration by removing the `default` values, so this PR is putting them back the way they were.

sbordet commented on jetty-server/src/main/config/etc/jetty-gzip.xml:
```diff
@@ -13,11 +13,7 @@
-        <Set name="compressionLevel" property="jetty.gzip.compressionLevel"/>
         <Set name="inflateBufferSize" property="jetty.gzip.inflateBufferSize"/>
-        <Set name="inflaterPoolCapacity" property="jetty.gzip.inflaterPoolCapacity"/>
-        <Set name="deflaterPoolCapacity" property="jetty.gzip.deflaterPoolCapacity"/>
```
I think this XML is out-of-date. It should be changed to allocate the `InflaterPool` and `DeflaterPool` and the relevant properties set on the pools (rather than on the handler as it is done here).

janbartel replied:
Is it fine to take a all the default values for the `InflaterPool` and `DeflaterPool`?

gregw replied:
I think yes.

Note that I am going to open another issue.  We need a way to check that `property` attributes actually exist, else we can change the code and not notice the XML is wrong unless the property is actually set.     I think we should always validate that there is at least one matching setter, even if the property is not set.

janbartel replied:
@sbordet  I have configured an `InflaterPool` and a `DeflaterPool`.  Note that I have had to change the names of the properties in the `.mod` file as they no longer make sense - I've put them in the `jetty-rewrite.xml ` as `deprecated`. I've used the default values that are present in the code.

sbordet replied:
You did not put them in `jetty-rewrite.xml`, did you?

janbartel replied:
Sorry, working on too many things at once. They are of course in` jetty-gzip.xml`.

sbordet commented on jetty-server/src/main/config/modules/server.mod:
```diff
@@ -63,7 +63,7 @@ etc/jetty.xml
 
 ## URI Compliance: SAFE, STRICT
-# jetty.httpConfig.uriCompliance=SAFE
+# jetty.httpConfig.uriCompliance=DEFAULT
```
Update comment with possible values, as this may be included in documentation.

janbartel replied:
Done.

sbordet reviewed (changes requested):
There seems to be a lot of breaking changes in XML files that seem to be out of scope for this PR, if the scope was initially to fix `stopOnShutdown` behavior.

Also, feels like many/most properties should be `property` attributes so that embedded and standalone have the same configuration.
However, some `property` attribute was changed to `Property` element and viceversa, but could not find a clear logic for that?

janbartel committed:
Issue #6329 Change comment in .mod file

Signed-off-by: Jan Bartel <janb@webtide.com>

janbartel committed:
Issue #6329 Remove defaults from session-store-gcloud.xml

janbartel committed:
Issue #6329 Config InflaterPool and DeflaterPool

sbordet commented:
I think it's a mistake to try to revert 343cf73.

All releases of Jetty 10, since 10.0.0-alpha0 included, had the 343cf73 behavior.
Which means that no official release of Jetty 10 had any different behavior.

This PR is trying to revert that commit and instate many new behaviors, breaking who was relying on the behavior of basically all the Jetty 10 releases.

@janbartel if you look at the issue that generated this PR (#6329), it was about `jetty.server.stopAtShutdown` only.
That it was caused by 343cf73 does not mean that we have to change many other behaviors.

I'm fine to change `<Property>` elements into attributes or viceversa, as long as the behavior is maintained or it's a blatant bug like `stopAtShutdown`.

If there are wrong behaviors, let's review them in different issues so we can enumerate the behaviors that were changed one by one in the release notes, and at least users will know what behavior changed and why.

sbordet commented on jetty-server/src/main/config/etc/jetty-gzip.xml:
```diff
@@ -26,6 +22,19 @@
+        <Set name="deflaterPool">
+            <New class="org.eclipse.jetty.util.compression.DeflaterPool">
+                <Arg type="int"><Property name="jetty.gzip.deflaterPool.capacity" deprecated="jetty.gzip.deflaterPoolCapacity" default="1024"/></Arg>
+                <Arg type="int"><Property name="jetty.gzip.deflaterPool.compression" deprecated="jetty.gzip.compression" default="-1"/></Arg>
```
The `deprecated` attribute should be `jetty.gzip.compressionLevel`.
I would use `jetty.gzip.deflaterPool.compression*Level*` for the new name, otherwise looks like a boolean.

janbartel replied:
OK, fixed.

sbordet commented on jetty-server/src/main/config/modules/gzip.mod:
```diff
@@ -21,17 +21,23 @@ etc/jetty-gzip.xml
-## Inflater pool max size (-1 for unlimited, 0 for no pooling)
-# jetty.gzip.inflaterPoolCapacity=-1
+## Gzip compression level (-1 for default)
+# jetty.gzip.deflaterPool.compressionLevel=-1
```
This is the right property name, but the correspondent XML was using the wrong name.

janbartel replied:
Fixed.

gregw commented:
@sbordet you can't both argue that this is an issue that can be punted from a release train AND that the behavior should be kept as the changed one because of previous releases.

I think version 1 if this PR did just fix the missing attribute issue until @joakime pointed out the inadvertent behavior change.

I see no reason to keep any accidental behavior changes

janbartel committed:
Issue #6329 Fix gzip compressionLevel property

Signed-off-by: Jan Bartel <janb@webtide.com>

janbartel commented:
> I think it's a mistake to try to revert 343cf73.
> 
> All releases of Jetty 10, since 10.0.0-alpha0 included, had the 343cf73 behavior.
> Which means that no official release of Jetty 10 had any different behavior.

Which means that all official releases of jetty-10 are in fact incorrect/broken.
Just because we have had n broken releases is not a reason to make n+1.

> 
> This PR is trying to revert that commit and instate many new behaviors, breaking who was relying on the behavior of basically all the Jetty 10 releases.
> 
> @janbartel if you look at the issue that generated this PR (#6329), it was about jetty.server.stopAtShutdown only.
> That it was caused by 343cf73 does not mean that we have to change many other behaviors.
> 
> I'm fine to change <Property> elements into attributes or viceversa, as long as the behavior is maintained or it's a blatant bug like stopAtShutdown.
Well, the `InflaterPool/DeflaterPool `is a blatant bug. If anyone had actually configured a property value, jetty would not have been able to start.

The jetty-threadlimit.xml is also a blatant bug: if you enabled the threadlimit module, you would be forgiven for assuming it was actually limiting, but without this change it is not!

The only nice-to-haves is the clean-up of the session modules default values, the rest you can easily argue are blatant bugs, or rotten `.mod` files.

> If there are wrong behaviors, let's review them in different issues so we can enumerate the behaviors that were changed one by one in the release notes, and at least users will know what behavior changed and why.
